Docker Network UX & remote API changes

mavenugo opened the pull request:
This PR is a follow up for various discussion around docker network experimental UX and API.
Includes : 
- Moved the UX and API handling from docker/libnetwork to docker/docker
- Removed the experimental services concept & replaced with a simplified network UX and API.
- Replaced experimental `--publish-service` with `--net=<user-defined-network>`
- Moving all the relevant Network UX and Remote APIs out of experimental
- Removed experimental `--default-network` daemon flag
- Neccessary backend changes to accomodate multiple networks connect UX per container
- Integration Tests
- Initial Docs update

Note that this PR updates the UX and API aspects of the CNM functionality.
there are a few related back-end issues still open (https://github.com/docker/libnetwork/issues/567, https://github.com/docker/libnetwork/issues/570, https://github.com/docker/libnetwork/issues/568) and will be addressed as part of libnetwork vendor-in in subsequent PRs.

UX (https://github.com/mavenugo/docker/blob/ux/docs/userguide/dockernetworks.md) 
API (https://github.com/mavenugo/docker/blob/ux/docs/reference/api/docker_remote_api_v1.21.md#25-networks).

mavenugo committed:
Network remote APIs using new router, --net=<user-defined-network> changes

* Moving Network Remote APIs out of experimental
* --net can now accept user created networks using network drivers/plugins
* Removed the experimental services concept and --default-network option
* Neccessary backend changes to accomodate multiple networks per container
* Integration Tests

Signed-off-by: David Calavera <david.calavera@gmail.com>
Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
Network UX and integration tests

* Exiting experimental network UX
* removed experimental service UX
* integrated with the new network remote API

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
Networking API and UX documentation

More doc updates will follow

Signed-off-by: Madhu Venugopal <madhu@docker.com>

cpuguy83 commented:
Some notes from playing with this:

1. Disconnecting a container from a network does not seem to remove /etc/hosts entries from the specified container for that network, though it seems the other containers on the network are updated to remove the container's entry

2. Docs say that you can only connect a running container to a network, but running or not seems to be ok (which I think is good).  Initially from the docs I thought it would be impossible to pre-setup a all of a container's network needs (ie, when it needs to be on more than one network) prior to starting it... although I did notice that if you have a stopped container, connect it to a network, then start it, you get a warning: `Could not connect container fdbfcb3fb78ba0541accf1e77e8df9168c09f0fe0e9cac550d52fff50ba0276e :a sandbox has already joined the endpoint`, even though it does work.

3. It is currently possible to remove a network which is in use, given that the container(s) using that network are also on another network.

mavenugo commented:
@cpuguy83 thanks for the feedback. Will look into these.

mavenugo commented:
@cpuguy83 #2 and #3 should be resolved now with the updated check for active containers. As discussed in IRC, we are trying to be consistent with the current network resource usage when the container is down and hence, its not appropriate to hold onto network resources when it is not active. But having said that, there has been requests for ip-address stability across container restarts (https://github.com/docker/libnetwork/issues/569). When we resolve that, we could potentially remove this restriction and allow users to connect and disconnect from networks even when the container is down.

mavenugo commented:
The windows CI failure is not related to this PR and I see multiple PRs failing on the same issue.

tombee commented:
Hey @mavenugo, awesome work on this -- this is exciting stuff.  The simplified UX is nice! :+1: 

I've been trying to get it working, but I'm finding that my overlay network only works on a single host and not multi-host.  It seems that the ad

```
tom at debian in ~
$ docker-machine create \
→     -d virtualbox \
→     consul
Creating VirtualBox VM...
Creating SSH key...
Starting VirtualBox VM...
Starting VM...
To see how to connect Docker to this machine, run: docker-machine env consul

tom at debian in ~
$ docker $(docker-machine config consul) run -d \
→     -p "8500:8500" \
→     -h "consul" \
→     progrium/consul -server -bootstrap
Unable to find image 'progrium/consul:latest' locally
latest: Pulling from progrium/consul
31f630c65071: Pull complete 
cadd1fc80511: Pull complete 
617a0e174acf: Pull complete 
8142b9305a34: Pull complete 
8e614066f6a6: Pull complete 
8ba0a056d1e3: Pull complete 
d79f17b5cf8f: Pull complete 
1222ac49a0eb: Pull complete 
8d3315362a90: Pull complete 
8359e65c65c1: Pull complete 
53d68e7edf50: Pull complete 
e57fb0e987c9: Pull complete 
39dd286a1371: Pull complete 
0ed95048b811: Pull complete 
d9333eedb08a: Pull complete 
18d2154179cc: Pull complete 
ff1cf6216ab6: Pull complete 
e66fb6787628: Pull complete 
Digest: sha256:8cc8023462905929df9a79ff67ee435a36848ce7a10f18d6d0faba9306b97274
Status: Downloaded newer image for progrium/consul:latest
6d4f2090880e0a54965137960a33da041effc8916ff2d7616853647a44150515

tom at debian in ~
$ docker-machine create \
→     -d virtualbox \
→     --swarm \
→     --swarm-master \
→     --swarm-discovery="consul://$(docker-machine ip consul):8500" \
→     --engine-opt="cluster-store=consul://$(docker-machine ip consul):8500" \
→     --engine-label="com.docker.network.driver.overlay.bind_interface=eth0" \
→     swarm-0
Creating VirtualBox VM...
Creating SSH key...
Starting VirtualBox VM...
Starting VM...
To see how to connect Docker to this machine, run: docker-machine env swarm-0

tom at debian in ~
$ docker-machine create \
→     -d virtualbox \
→     --swarm \
→     --swarm-discovery="consul://$(docker-machine ip consul):8500" \
→     --engine-opt="cluster-store=consul://$(docker-machine ip consul):8500" \
→     --engine-label="com.docker.network.driver.overlay.bind_interface=eth0" \
→     swarm-1
Creating VirtualBox VM...
Creating SSH key...
Starting VirtualBox VM...
Starting VM...
To see how to connect Docker to this machine, run: docker-machine env swarm-1

tom at debian in ~
$ dockerx $(docker-machine config swarm-0) \
→     network create -d overlay demo
b493d420c07f713ddad7324d00ecac7fba87edfcdc332460bb6fc43efcace87e

tom at debian in ~
$ dockerx $(docker-machine config --swarm swarm-0) \
→     run -d --name nginx \
→     --env="constraint:node==swarm-0" \
→     --net="demo" \
→     --publish="80:80" \
→     nginx
67f87b75cbba7204d8d26d2764c31520479f9a71adb0873063d2e71bd3d65c44

tom at debian in ~
$ dockerx $(docker-machine config --swarm swarm-0) \
→     run -it --rm \
→     --env="constraint:node==swarm-0" \
→     --net="demo" \
→     debian ping nginx
PING nginx (172.21.0.1): 56 data bytes
64 bytes from 172.21.0.1: icmp_seq=0 ttl=64 time=0.219 ms
64 bytes from 172.21.0.1: icmp_seq=1 ttl=64 time=0.049 ms
64 bytes from 172.21.0.1: icmp_seq=2 ttl=64 time=0.055 ms
^C--- nginx ping statistics ---
10 packets transmitted, 10 packets received, 0% packet loss
round-trip min/avg/max/stddev = 0.049/0.070/0.219/0.050 ms

tom at debian in ~
$ dockerx $(docker-machine config --swarm swarm-0) \
→     run -it --rm \
→     --env="constraint:node==swarm-1" \
→     --net="demo" \
→     debian ping nginx
PING nginx (172.21.0.1): 56 data bytes
92 bytes from 12cb9f8fa0d1 (172.21.0.2): Destination Host Unreachable
^C--- nginx ping statistics ---
12 packets transmitted, 0 packets received, 100% packet loss
```

Here's a dump of the in-container `/etc/hosts` files:
```
$ dockerx $(docker-machine config --swarm swarm-0)     run -it --rm     --env="constraint:node==swarm-1"     --net="demo"     debian cat /etc/hosts
172.21.0.2	16b43b2abfaa
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.21.0.1	nginx
172.21.0.1	nginx.demo
172.21.0.2	romantic_khorana
172.21.0.2	romantic_khorana.demo
172.18.0.2	gateway_16b43b2abfaa
172.18.0.2	gateway_16b43b2abfaa.docker_gwbridge

$ dockerx $(docker-machine config --swarm swarm-0)     run -it --rm     --env="constraint:node==swarm-0"     --net="demo"     debian cat /etc/hosts
172.21.0.2	ff3187a2922d
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.21.0.1	nginx
172.21.0.1	nginx.demo
172.21.0.2	prickly_mccarthy
172.21.0.2	prickly_mccarthy.demo
172.18.0.3	gateway_ff3187a2922d.docker_gwbridge
172.18.0.1	gateway_67f87b75cbba
172.18.0.1	gateway_67f87b75cbba.docker_gwbridge
172.18.0.3	gateway_ff3187a2922d
```

mavenugo commented:
@tombee thanks for the feedback. I see that you are missing the 
`--engine-label="com.docker.network.driver.overlay.neighbor_ip=$(docker-machine ip swarm-0)" \`
on the swarm node (as mentioned in https://github.com/docker/docker/blob/master/experimental/compose_swarm_networking.md).

Please note that these engine labels will go away once libnetwork integration with docker discovery is complete.

tombee commented:
@mavenugo Ahh!  Sorry, misunderstood -- thought that label was gone away already.  Nice :)

tombee commented:
Okay, this is my latest attempt at running this multi-host, I'm seeing this error:
```
$ dockerx $(docker-machine config --swarm swarm-0) \
    run -it --rm \
    --env="constraint:node==swarm-1" \
    --net="demo" \
    busybox ping nginx

Error response from daemon: Post https://192.168.99.102:2376/v1.21/containers/0f159d19b668a8370757264d2b0d3e779f381b1a4f3a7460e96bc9cf1fe7ca22/start: EOF
```

These are the errors in the log:
```
time="2015-09-29T22:57:51.946531565Z" level=info msg="GET /v1.15/containers/json?all=1&size=0&filters={\"id\":[\"0f159d19b668a8370757264d2b0d3e779f381b1a4f3a7460e96bc9cf1fe7ca22\"]}" 
time="2015-09-29T22:57:51.970614753Z" level=debug msg="Calling POST /containers/{name:.*}/start" 
time="2015-09-29T22:57:51.970661162Z" level=info msg="POST /v1.21/containers/0f159d19b668a8370757264d2b0d3e779f381b1a4f3a7460e96bc9cf1fe7ca22/start" 
time="2015-09-29T22:57:51.976788212Z" level=warning msg="Failed adding service host entries to the running container: open /mnt/sda1/var/lib/docker/containers/f1d9543e0862ed6a349c0224423042018963e18ed238e21573e82921fe8ba674/hosts: no such file or directory" 
time="2015-09-29T22:57:51.976840810Z" level=warning msg="Failed adding service host entries to the running container: open /mnt/sda1/var/lib/docker/containers/fba80a337917cb716a7b1d4038fa1601de9ce0b35bb24003b0f8b5c0999a3a77/hosts: no such file or directory" 
time="2015-09-29T22:57:51.976871637Z" level=warning msg="Failed adding service host entries to the running container: open /mnt/sda1/var/lib/docker/containers/5fa12bd3bcbc1764570e39a32b3a691bb468fdf2567a19272f0d89eccfc9d951/hosts: no such file or directory" 
2015/09/29 22:57:51 http: panic serving 192.168.99.101:48995: runtime error: invalid memory address or nil pointer dereference
goroutine 584 [running]:
net/http.func·011()
	/usr/local/go/src/net/http/server.go:1130 +0xbb
github.com/hashicorp/serf/serf.(*Serf).LocalMember(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/hashicorp/serf/serf/serf.go:689 +0xec
github.com/docker/libnetwork/drivers/overlay.(*driver).Join(0xc2081b0960, 0xc2082c6380, 0x40, 0xc20850a8c0, 0x40, 0xc20849a240, 0x22, 0x7f438a222870, 0xc208772a20, 0x0, ...)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/joinleave.go:76 +0xde8
github.com/docker/libnetwork.(*endpoint).sbJoin(0xc208772a20, 0x7f438a222810, 0xc208be1200, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/endpoint.go:241 +0x4fc
github.com/docker/libnetwork.(*endpoint).Join(0xc208772a20, 0x7f438a222810, 0xc208be1200, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/endpoint.go:205 +0x1a8
github.com/docker/docker/daemon.(*Container).connectToNetwork(0xc208946200, 0x7f438a211798, 0xc208bf17d0, 0xc208240e90, 0x4, 0xc208bd3001, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/container_unix.go:967 +0x782
github.com/docker/docker/daemon.(*Container).allocateNetwork(0xc208946200, 0x7f438a211798, 0xc208bf17d0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/container_unix.go:880 +0x373
github.com/docker/docker/daemon.(*Container).initializeNetworking(0xc208946200, 0x7f438a211798, 0xc208bf17d0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/container_unix.go:1042 +0x4ee
github.com/docker/docker/daemon.(*Container).Start(0xc208946200, 0x7f438a211798, 0xc208bf17d0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/container.go:278 +0x27c
github.com/docker/docker/daemon.(*Daemon).ContainerStart(0xc208069800, 0x7f438a211798, 0xc208bf17d0, 0xc208bf4557, 0x40, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/start.go:48 +0x29f
github.com/docker/docker/api/server.(*Server).postContainersStart(0xc2081ce940, 0x7f438a211798, 0xc208bf17d0, 0x7f438a211630, 0xc20859ef00, 0xc2085360d0, 0xc208bf1440, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/container.go:186 +0x176
github.com/docker/docker/api/server.*Server.(github.com/docker/docker/api/server.postContainersStart)·fm(0x7f438a211798, 0xc208bf17d0, 0x7f438a211630, 0xc20859ef00, 0xc2085360d0, 0xc208bf1440, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/server.go:369 +0x78
github.com/docker/docker/api/server.func·010(0x7f438a211798, 0xc208bf17d0, 0x7f438a211630, 0xc20859ef00, 0xc2085360d0, 0xc208bf1440, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/middleware.go:97 +0x6ab
github.com/docker/docker/api/server.func·009(0x7f438a211798, 0xc208bf1680, 0x7f438a211630, 0xc20859ef00, 0xc2085360d0, 0xc208bf1440, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/middleware.go:76 +0x10e
github.com/docker/docker/api/server.func·008(0x7f438a211798, 0xc208bf1680, 0x7f438a211630, 0xc20859ef00, 0xc2085360d0, 0xc208bf1440, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/middleware.go:59 +0x479
github.com/docker/docker/api/server.func·006(0x7f438a211798, 0xc208bf1680, 0x7f438a211630, 0xc20859ef00, 0xc2085360d0, 0xc208bf1440, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/middleware.go:27 +0x1d9
github.com/docker/docker/api/server.func·007(0x7f438a211798, 0xc208bf1680, 0x7f438a211630, 0xc20859ef00, 0xc2085360d0, 0xc208bf1440, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/middleware.go:37 +0x1e9
github.com/docker/docker/api/server.func·014(0x7f438a211630, 0xc20859ef00, 0xc2085360d0)
	/go/src/github.com/docker/docker/api/server/server.go:312 +0x3c4
net/http.HandlerFunc.ServeHTTP(0xc208244840, 0x7f438a211630, 0xc20859ef00, 0xc2085360d0)
	/usr/local/go/src/net/http/server.go:1265 +0x41
github.com/gorilla/mux.(*Router).ServeHTTP(0xc2081cb810, 0x7f438a211630, 0xc20859ef00, 0xc2085360d0)
	/go/src/github.com/docker/docker/vendor/src/github.com/gorilla/mux/mux.go:98 +0x297
net/http.serverHandler.ServeHTTP(0xc2081c4300, 0x7f438a211630, 0xc20859ef00, 0xc2085360d0)
	/usr/local/go/src/net/http/server.go:1703 +0x19a
net/http.(*conn).serve(0xc2084d66e0)
	/usr/local/go/src/net/http/server.go:1204 +0xb57
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1751 +0x35e
time="2015-09-29T22:57:51.997554320Z" level=debug msg="Calling DELETE /containers/{name:.*}"
```

These are the steps I'm using:
```
docker-machine create \
    -d virtualbox \
    consul

docker $(docker-machine config consul) run -d \
    -p "8500:8500" \
    -h "consul" \
    progrium/consul -server -bootstrap
                                                                                                   
docker-machine create \
    -d virtualbox \
    --swarm \
    --swarm-master \
    --swarm-discovery="consul://$(docker-machine ip consul):8500" \
    --engine-opt="cluster-store=consul://$(docker-machine ip consul):8500" \
    --engine-label="com.docker.network.driver.overlay.bind_interface=eth0" \
    swarm-0

docker-machine create \
    -d virtualbox \
    --swarm \
    --swarm-discovery="consul://$(docker-machine ip consul):8500" \
    --engine-opt="cluster-store=consul://$(docker-machine ip consul):8500" \
    --engine-label="com.docker.network.driver.overlay.bind_interface=eth0" \
    --engine-label="com.docker.network.driver.overlay.neighbor_ip=$(docker-machine ip swarm-0)" \
    swarm-1

dockerx $(docker-machine config swarm-0) \
    network create -d overlay demo

dockerx $(docker-machine config --swarm swarm-0) \
    run -d --name nginx \
    --env="constraint:node==swarm-0" \
    --net="demo" \
    --publish="80:80" \
    nginx

dockerx $(docker-machine config --swarm swarm-0) \
    run -it --rm \
    --env="constraint:node==swarm-1" \
    --net="demo" \
    busybox ping nginx
```

mavenugo commented:
@tombee thanks for the feedback and trying it out using the multi-host networking feature.

ibuildthecloud commented:
On `docker network ls` both `-l` and `-n=` don't seem to do anything.

ibuildthecloud commented:
There is no `docker network inspect foo`

ibuildthecloud commented:
Sorry, just deleted some comment, I realized I was testing the wrong branch... too many PRs :)

ibuildthecloud commented:
At this point I don't see the value in `docker network inspect`.  With the absence of the `-f` flag like `docker inspect -f '{{...}'`, it just returns the same info in `docker network ls`

ibuildthecloud commented:
`docker network ls -l` and `docker network ls -n=1` don't seem to work

ibuildthecloud commented:
I stand corrected.  I see that `docker network inspect` will return the container info.  I think that `containers` should be `containers: {}` if there are no containers in the network.  Not just drop the field.  So omitempty shouldn't be set.

Also, why don't stopped container show up in `network inspect`?

ibuildthecloud commented:
I think its critical to return the scope of the network in `/networks` and in the inspect.  It will be important to understand if the network is local or global, especial for orchestration systems.

ibuildthecloud commented:
In the container inspect why would we prefer to put the network name in `NetworkSettings.Networks` as opposed to the ID.  It would seem to me that with local and global networks it would be possible to get a name conflict.  For example, host A has local net called `foo` and then host B creates global net called `foo`.

mavenugo commented:
@ibuildthecloud thanks for trying out the changes. 

> docker network ls -l and docker network ls -n=1 don't seem to work
Will fix it

> So omitempty shouldn't be set.
Agreed. will fix it

> why don't stopped container show up in network inspect?
Thats because, with the current backend implementation, When a container is stopped, we cleanup the endpoints and free up sandbox, endpoint and cleanup all the network resources allocated for that. These are operational state. If the container is started again, we will make use of the NetworkSettings (Configuration state) to recreate the resources appropriately.
Since the network resources are freed up when the container is stopped, inspect just doesnt show up any information for that container. Also, if all the containers in a network is stopped, user can even remove the network.
Having said that, if we support https://github.com/docker/libnetwork/issues/569, then this will change and the network resources will be retained even if the container is stopped.

> I think its critical to return the scope of the network in /networks
okay. That requires a back-end API change. am trying to keep this PR purely UX centric (as you can see there is no libnetwork vendor-in in this PR) to ease the reviewer burden. We can update it in another PR after the backend is updated.

> why would we prefer to put the network name in NetworkSettings.Networks as opposed to the ID
The main purpose is that, we dont persist the default networks (bridge, none and host). Hence upon restart these default networks get a new ID, but the name remains the same. That is the only reason. 
The question on why we dont persist the default networks is a long conversation. But, I prefer storing the ID as well. If you look at the code carefully, I made sure the []networks can take in both name or id. Because, when the backend changes (to accomodate default network persistence), we can flip the switch as easily to start storing id. Infact we can do it for 1.9 after the libnetwork backend supports default network persistence.

cpuguy83 commented:
Design :+1: I like this a lot.

Only thing that was a little weird is the ordering of `docker network connect <network> <container>`, I naturally want to switch `<container>` and `<network>`... just throwing it out there in case anyone else did this as well... but as is, is ok.

mavenugo commented:
thanks @cpuguy83

> the ordering of docker network connect <network> <container>, I naturally want to switch <container> and <network>

Yes, we did think of that. but, decided to settle with <network> <configuration> due to a matter of convention where folks would come to expect the <network> param to be the key for all the `docker network` command. Hence placing it first seems more appropriate.

ibuildthecloud commented:
@mavenugo I think IP stability is important.  Specifically I'm dealing with hadoop right now and if you change the IP bad things happen.  We are currently using our managed networking which maintains IPs across reboot, but I think that should be a core feature.

For the NetworkSettings.Networks, is it possible there is a compromise such that IDs go in there except host, bridge, none put the name.

I really need to know if the networks are global vs local, can that be added to the API?

Last thing.  I noticed that `docker volume` defaults to `docker volume ls` where as `docker network` prints the help.  I think between @cpuguy83 and @mavenugo we should have a consistent approach.

cpuguy83 commented:
@ibuildthecloud IP stability is completely separate from this change, especially since there is no stability today.

ibuildthecloud commented:
@cpuguy83 Agreed, on the IP stability part, sorry for the tangent.

mavenugo commented:
@ibuildthecloud yes. we will get both Scope and NetworkSettings to carry a list of IDs. As I explained both of these requires backend changes. Also, the NB-API and UX is not frozen. We are still to add the ipam-driver details to the network once IPAM driver changes are merged in libnetwork.

I would like to keep this PR clean and isolated to address the UX/API in daemon without the need to vendor in libnetwork backend changes. Hope that make sense.

alfredr commented:
@cpuguy83's scenario happens for me and some others as well, except we're unable to start the container at all (see https://github.com/docker/libnetwork/issues/573).

mavenugo commented:
@alfredr these are unrelated issues. the one @cpuguy83 was seeing was due to an issue of network connect happening on a stopped container. By design, we dont support that today and it was a bug in this PR that I later resolved.
The one that you are facing is completely different and will be addressed as part of libnetwork backend changes.

alfredr commented:
@mavenugo. I see. Thank you

mavenugo commented:
thanks @calavera for rebasing this PR on top of the #16539.
Updated the PR with those changes. Once it turns green, I can squash the first 2 commits only if that makes sense for the reviewers. Else, I can leave the commits as is.

calavera commented:
Design looks good to me too. Much more simple that the current networking commands in experimental and more aligned with the volume commands.

I've moved it to code-review.

cpuguy83 commented:
A lot of the work being done at the API layer should be pushed down into daemon level functions, especially anything talking to the network controller such that the API layer is only ever interacting with the daemon, objects from `types/`, or built-in types.

tomdee commented:
This doesn't mention IPAM or IP addresses in general. Does that mean the IPAM work won't be in 1.9 stable?

mavenugo commented:
IPAM work is not complete in libnetwork. We are expecting it to be done soon. When that is censored in, we will update the UX as well.

tomdee commented:
I've just tried out this code and it seems to be working with Calico. Good work! Thanks @mavenugo 

The "Connecting to multiple networks" part isn't working so I'll need to keep thinking about that. I hit
```
$ docker-dev network connect calico-net2 cont11
Error response from daemon: failed to add interface tmpe17ab68e9c1 to sandbox: error setting interface "tmpe17ab68e9c1" routes to ["172.17.8.10/32"]: file exists
```

mavenugo commented:
@tombee thanks for trying it out with your network plugin.
As you know, this is purely an UX change and it uses exactly the same APIs that `service attach` and `service detach` was using. if there is any issue, it should be in the back-end. Pls let me know if you find anything specific around Calico.

mapuri commented on api/server/router/network/network.go:
```diff
@@ -1,26 +1,50 @@
+		local.NewGetRoute("/networks/{id:.*}", r.getNetwork),
+		// POST
+		local.NewPostRoute("/networks/create", r.postNetworkCreate),
+		local.NewPostRoute("/networks/{id:.*}/connect", r.postNetworkConnect),
```
minor:

since this post is going to connect a `container` to a `network`, it might be better to have the container-id in the url itself i.e. 

```
"/networks/{id:.*}/connect/{id:.*}",
```

same will be true for `disconnect`

mavenugo replied:
its highly likely that we will expand the connect data to more than container-id (such as alias and other endpoint specific data). When it comes, we dont want to change the API again. So, I prefer to keep the current structure and it can easily be expanded by having backward compatible clients.

mapuri commented:
@mavenugo +1 for the new much more explicit UX!

I might have missed this but looks like for the container to join multiple networks the user needs to issue a  `connect` after it is created. Just wondering if you also considered allowing `--net` flag to be specified multiple times during `docker run`? That would allow the container to be put in all networks at it's creation than adding it later and might be a preferred option in certain situations.

mavenugo commented:
@mapuri if we have multiple --net in run, how can the user map the network with other endpoint specific configurations (such as mac-address, ip-address (in the future), port-expose etc...) ?
This PR was raised to simplify the User experience and I dont want to complicate it by adding more knobs that are unusable.

mapuri commented:
@mavenugo 
> how can the user map the network with other endpoint specific configurations (such as mac-address, ip-address (in the future), port-expose etc...) ?

good point, yes if we try to map to existing flags then it will deter the user experience. The current PR is a good starting point for now.

But I think we might still need to address a case of the container that can belong to multiple networks at it's start.

I think one option might be to introduce a separate UX for endpoint creation where such network specific properties can be specified at endpoint level as generic cli flags. 

Alternatively, the network specific options can be passed as options (or labels) to the driver. I think this will also depend on how https://github.com/docker/libnetwork/pull/222 get's integrated in the UX and can wait until that happens.

mavenugo commented:
@mapuri I don't think we can have this requirement of multiple networks addressed in this PR for 1.9.
With all the discussions on UX in the past, if we digress from what we have now, then it is a good possibility that we may not get the basic UX for 1.9. If you still want to pursue such an option, I would recommend you to push another PR.

Having said that, a container that is attached to multiple networks (using `network connect` command) can be restarted and it will be connected to all these networks automatically with the same effect that you are looking for. But ofcourse, that would require a container restart after `network connect`.
Another possibility is the support to have `network connect` working on stopped container. That has other implications as discussed earlier in the thread wrt resource utilization even when the container is not active.
In any case, I think we should take it up as a separate discussion and use this PR to get a simplified UX working and exit experimental for 1.9.

mapuri commented:
>  I think we should take it up as a separate discussion and use this PR to get a simplified UX working and exit experimental for 1.9.

yes, sgtm !

mavenugo commented:
@cpuguy83 the reason I decided to expose networkcontroller to the api layer is that, unlike other UX, all the network centric UX will end up being processed only in libnetwork. If we hide the networkcontroller out of api and into daemon, then we will have API duplication at the daemon level (like, there will be atleast 1 API per REST duplicated between docker daemon and libnetwork). I was just trying to avoid that by exposing the network controller to API layer and it calling libnetwork directly. In the future as we expand the scope of the UX for more networking functionality, the same trend of API duplication will continue which I thought is artificial. WDYT ?

cpuguy83 commented:
@mavenugo I believe the idea is that we don't want to expose internal data structures to the API so that we can actually extract the API into a library... but someone can correct me here if I'm wrong (or confirm).

thockin commented:
@mavenugo This is a MUCH nicer UX.  It feels like a clean layer, now.  Thanks.

Is there a way to opt out of the /etc/hosts management?  Given that we have our own DNS, the names that are auto-assigned are not so useful for us.

mavenugo commented:
@thockin thanks. as you understand, this is a UX only PR and it is not replacing any of the existing functionality. am resisting any backend functionality changes in this PR. We can certainly look into #14633 once this is merged.

thockin commented:
I think that is the wrong issue number :)

On Thu, Oct 1, 2015 at 5:52 PM, Madhu Venugopal <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> thanks. as you understand, this is
> a UX only PR and it is not replacing any of the existing functionality. am
> resisting any backend functionality changes in this PR. We can certainly
> look into #14633 <https://github.com/docker/docker/issues/14633> once
> this is merged.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16645#issuecomment-144889614>.
>

mavenugo commented:
@thockin :) okay, just put the wrong 2 and 2 together. but i hope you got the point and agree with the rationale.

thockin commented:
Oh, sure.  I just read the docs on this PR, not the PR itself.  I just want
you all think about opt-out of DNS when you;re doing the real work :)

Appreciate it!


On Thu, Oct 1, 2015 at 5:57 PM, Madhu Venugopal <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> :) okay, just put the wrong 2 and 2
> together. but i hope you got the point and agree with the rationale.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16645#issuecomment-144890481>.
>

aanand commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -2473,6 +2473,218 @@ Status Codes
+  {
+    "Id": "22be93d5babb089c5aab8dbc369042fad48ff791584ca2da2100db837a1c7c30",
+    "Warning": ""
+  }
```
Testing this PR locally, I get lowercase field names here (`id` and `warning`, not `Id` and `Warning`).

mavenugo replied:
yes. it should be lowercase. will fix it.

aanand commented on docs/userguide/dockernetworks.md:
````diff
@@ -0,0 +1,511 @@
+
+$ sudo docker run -itd --name=container2 busybox
+bda12f8922785d1f160be70736f26c1e331ab8aaf8ed8d56728508f2e2fd4727
+```
````
This code snippet doesn't demonstrate statement that precedes it. I think it'd be better to combine it with the `inspect` command below, which *does* demonstrate that the containers have joined the `bridge` network.

mavenugo replied:
okay. will do.

aanand commented:
One dramatic change here is that I can seemingly no longer choose the name by which a container joins a network.

If that's the case, this is going to be a serious issue for Compose: services have names like `db`, but their containers have names like `myproject_db_1`. It really wouldn't be OK to expect users to use that as a hostname when connecting to their database. Have I misunderstood?

mavenugo commented:
@aanand the containers connect to the network with their `--name` as the dns name with which other containers reach it. Since the service concept in engine is being reconsidered, we need to find another way to provide a flexible way to assign name aliases for a container. I was thinking of another option like `--connect-as`, when used will act as an alias for the container in that network. Compose can make use of it as the `service` name and also it can act as a link replacement with the alias feature. WDYT ?
Since this is a mandatory requirement for Compose integration, we can take this discussion in a subsequent PR.

aanand commented:
@mavenugo Sounds good to me!

icecrime commented on docs/userguide/dockernetworks.md:
```diff
@@ -0,0 +1,511 @@
+$ sudo docker network ls
+NETWORK ID          NAME                DRIVER
+7fca4eb8c647        bridge              bridge
+9f904ee27bf5        none                null
```
Small nit: driver `none` would read better than `null` ;-)

mavenugo replied:
okay. thats a backend change, and the driver must register itself as none and it will be done :).
Not in this PR though.

mavenugo replied:
@icecrime driver with the name `null` seems to be correct. a driver marked `none` means that there is no driver backing the network named `none`. But it is not true it does have its driver named `null` with its only job is to provide `loopback` interface to the container. So, I guess we should leave this as is. Unless you have other ideas.

icecrime commented on docs/userguide/dockernetworks.md:
```diff
@@ -0,0 +1,511 @@
+ff02::1 ip6-allnodes
+ff02::2 ip6-allrouters
+172.17.0.2      container1
+172.17.0.2      container1.bridge
```
Sorry for the possible dump question: why the `<name>.bridge` entry, and why is is equal to `<name>`?

mavenugo replied:
since a container can be attached to multiple networks and <name>.<network> provides a fully qualified name for a container. While the <name> is more for simplicity and when the user doesnt care about exact domain that the container belong to.

icecrime replied:
Ah got it! :+1:

icecrime commented:
I need more time to test and review the code, but loving the new UX :+1:

icecrime commented on api/client/network.go:
```diff
@@ -1,14 +1,304 @@
+}
+
+var (
+	networkCommands = []command{
```
Is there a reason to not make this a map?

mavenugo replied:
no apparent reason. I will change it.

icecrime commented on api/client/network.go:
```diff
@@ -1,14 +1,304 @@
+	cmd := Cli.Subcmd("network", []string{"COMMAND [OPTIONS]"}, networkUsage(), false)
+	cmd.Require(flag.Min, 1)
+	err := cmd.ParseFlags(args, true)
+	if err == nil {
```
`if err != nil`?

mavenugo replied:
since `network` is a top level command with mandatory sub-commands, `ParseFlags` will handle the help usage if the user executes `docker network` without any argument.
But when the user executes an invalid argument such as `docker network foo`, `ParseFlags` cannot catch an error and hence it will return err as nil. But, since there is no corresponding `CmdNetworkFoo` handler, we must handle this condition under the `err == nil` condition.

But, it is infact, okay and generic to infact remove such a check all together and it is going to work fine.

icecrime commented on daemon/config_experimental.go:
```diff
@@ -1,9 +0,0 @@
-import flag "github.com/docker/docker/pkg/mflag"
-
-func (config *Config) attachExperimentalFlags(cmd *flag.FlagSet, usageFn func(string) string) {
-	cmd.StringVar(&config.DefaultNetwork, []string{"-default-network"}, "", usageFn("Set default network"))
```
There are no remaining option to set a daemon-wide default network?

vieux replied:
Same question, it was useful for swarm

mavenugo replied:
--default-network being a daemon option, and it is not synchronized across the cluster, it is quite easy to have different --default-networks in different nodes in the cluster. Hence, it is upto the user to make sure it is consistent. That results in unexpected behaviour when containers are launched in different nodes.
Also, what is the expectation if the default-network changes after a daemon restart ?
Should the stopped containers start using the new network as the default-network or stick to the older one ?

Now that swarm and compose will have native integration with the new networking features, it is appropriate for swarm to make sure it creates the networks it want to use before using it.

Thats the reason for the removal.

icecrime replied:
Do we intend to change the default to `overlay` in a later version?

mavenugo replied:
we could. also, we can still provide user configurable default-network once we have a proper daemon level state synchronization.

icecrime commented on api/types/types.go:
```diff
@@ -342,3 +342,39 @@ type VolumeCreateRequest struct {
+}
+
+// NetworkConnect represents a list of containers to be connected to a network
+type NetworkConnect struct {
```
Given it's used by both connect and disconnect, maybe rename to `NetworkConnectivityRequest`?

mavenugo replied:
yes. created `NetworkDisconnect`. am expecting `NetworkConnect` to expand further and hence keeping them independent is a good start.

icecrime replied:
:+1:

icecrime commented on api/client/network.go:
```diff
@@ -1,14 +1,304 @@
+func (cli *DockerCli) CmdNetworkLs(args ...string) error {
+	cmd := Cli.Subcmd("network ls", []string{""}, "Lists all the networks created by the user", false)
+	quiet := cmd.Bool([]string{"q", "-quiet"}, false, "Only display numeric IDs")
+	noTrunc := cmd.Bool([]string{"#notrunc", "-no-trunc"}, false, "Do not truncate the output")
```
Why support a deprecated `#notrunc` in the new code?

mavenugo replied:
Its legacy code from libnetwork being carried on. Will remove this.

icecrime commented on api/client/network.go:
```diff
@@ -1,14 +1,304 @@
+	if err := json.Indent(indented, obj, "", "    "); err != nil {
+		return err
+	}
+	if _, err := io.Copy(cli.out, indented); err != nil {
```
It's surprising to have this one dump raw JSON instead of a structured table.

mavenugo replied:
the network inspect output will certainly expand to accomodate more data in it (such as stats) & more network resources. Structured table doesn't suit well. `network inspect` (similar to `docker inspect`) will be highly appreciated by  operators and devs alike.

icecrime commented on api/client/network.go:
```diff
@@ -1,14 +1,304 @@
+// CmdNetworkInspect handles Network inspect UI
+func (cli *DockerCli) CmdNetworkInspect(args ...string) error {
+	cmd := Cli.Subcmd("network inspect", []string{"NETWORK"}, "Displays detailed information on a network", false)
+	cmd.Require(flag.Exact, 1)
```
Nice to have: `--format` on `docker inspect` is particularly useful, we should try to have the same here if it's not too much of an effort.

mavenugo replied:
Sure. the inspect output atm is simple. But with the introduction of ipam driver (https://github.com/docker/libnetwork/pull/525), we are adding more stuffs to the network inspect and that might make the need for --format useful. I think we can decide on that in the subsequent PR for the ipam UX changes. WDYT ?

icecrime replied:
I wouldn't block this PR on that as long as it is in the plans. What is the timeline for the IPAM UX changes?

mavenugo replied:
immediately after this PR and Discovery PRs are merged :)

icecrime commented on api/client/network.go:
```diff
@@ -1,14 +1,304 @@
+// Being a UI, its most likely that name will be used by the user, which is used to lookup
+// the corresponding ID. If ID is not found, this function will assume that the passed string
+// is an ID by itself.
+
```
:scissors:

mavenugo replied:
:) most of the UX code has moved from libnetwork to docker.
hence it has some legacy stuff in place. Will remove them once you are done with your initial round of reviews.

mavenugo replied:
`lookupNetworkID` will be entirely removed in the next update.

icecrime commented on api/client/network.go:
```diff
@@ -1,14 +1,304 @@
+// the corresponding ID. If ID is not found, this function will assume that the passed string
+// is an ID by itself.
+
+func lookupNetworkID(cli *DockerCli, nameID string) (string, error) {
```
Not sure about this one: up to 2 potential additional API calls for every possible network operation seems like a lot. Why not leave this to the daemon? As far as I can tell, this is how we do it for containers.

mavenugo replied:
this was done precisely to make the API, REST Compliant.
@aboch has some strong opinion in the RESTness of the APIs and hence we ended up with a REST-API 
that will accept just IDs and the complexity of name and ID is left to the client.

Now that we have moved the APIs into docker, the REST strictness doesnt apply.
@aboch are you okay if we change the opinion and implement the name vs ID resolution in the backend and
keep the client code simpler ?

icecrime commented on api/server/router/network/network_routes.go:
```diff
@@ -0,0 +1,248 @@
+				}
+				return false
+			}
+			n.netController.WalkNetworks(l)
```
I don't understand why we are not using this same `netController.WalkNetworks` function with a different parameter when we are filtering on name. This makes the two branches very different, and `findNetwork` happen to also deal with ids, so it's all quite confusing.

mavenugo replied:
reshuffled the entire UX & API logic to be more aligned with docker's handling of `name`, `partial-id` and `id` as part of API layer and not from the UX.
That took care of most of these comments.

icecrime commented on api/server/router/network/network_routes.go:
```diff
@@ -0,0 +1,248 @@
+		return c.NetworkByID(s)
+	case byName:
+		if s == "" {
+			s = c.Config().Daemon.DefaultNetwork
```
I'm confused: I can't find where the `DefaultNetwork` is set now that the command-line flag is gone

mavenugo replied:
DefaultNetwork is by default `bridge`. we provided an option to override it using the daemon flags.
But it is removed (https://github.com/docker/docker/pull/16645/files#r41079402) and hence this will always be "bridge", I kept the DefaultNetwork concept in the backend just to make sure we can reintroduce the UX if we can provide the consistency guarantees across the cluster in the future.

icecrime replied:
Sorry, I'm still not entirely sure I see where is `"bridge"` defined as the default: can you please point me to that line? I really just want to understand if it's in Docker or in libnetwork.

mavenugo replied:
FYI, in the later commits, findNetwork has been replaced by daemon.FindNetwork.

the default driver = bridge is identified in docker (https://github.com/docker/docker/blob/master/daemon/daemon_unix.go#L318) and passed to libnetwork as daemon option. libnetwork will make use of it as the default from there on.

icecrime commented on api/server/router/network/network_routes.go:
```diff
@@ -0,0 +1,248 @@
+	return r
+}
+
+func processCreateDefaults(c libnetwork.NetworkController, n *types.NetworkCreate) {
```
That function is a bit meh :-/  It's also used in a single place. Suggestion: can we rely on a `func newNetworkCreate() *types.NetworkCreate` to returns a default-initialized instance (with the `Driver` and options maps nicely instantiated), and just decode the JSON "over-it" and let it override whatever is specified as part of the JSON payload?

mavenugo replied:
yes. removed it entirely and made it as part of the backend logic.

icecrime commented on api/server/router/network/network_routes.go:
```diff
@@ -0,0 +1,248 @@
+
+func buildNetworkResource(nw libnetwork.Network) *types.NetworkResource {
+	r := &types.NetworkResource{}
+	if nw != nil {
```
That's a lot of nesting: can we rather return early?

icecrime commented on daemon/container_windows.go:
```diff
@@ -38,6 +39,16 @@ func (container *Container) initializeNetworking() error {
 }
 
+// ConnectToNetwork connects a container to the network
+func (container *Container) ConnectToNetwork(idOrName string) error {
```
Just making sure here: Windows containers still belong to a default network and have the same connectivity than today? We just don't benefit from "dynamic" network (dis)connections?

mavenugo replied:
WIndows containers have a different network model and it is not using the CNM model yet.
Hence dynamic network (dis)connection is not fully thought through. There is a separate effort
underway to bring windows closely aligned with CNM. this must be taken up along with that effort.

icecrime commented on daemon/container_unix.go:
```diff
@@ -1054,29 +1140,76 @@ func (container *Container) releaseNetwork() {
+		if err != nil {
+			continue
+		}
+		container.disconnectFromNetwork(n, false)
```
I don't understand what's wrong with having `disconnectFromNetwork` do `updateSettings` in all cases :-/ What makes this code path special?

mavenugo replied:
disconnectFromNetwork is a common function called from Configuration event (such as `docker network disconnect`) which needs to absolutely remove configuration states. And also, it is called from an operational event (such as `docker stop <container>`), which should not update the network configuration settings. When the container starts again, it must be automatically connected to the networks it was connected earlier through user configuration.
That makes this path special and hence requires special handling.

icecrime commented on daemon/container_unix.go:
```diff
@@ -920,6 +974,39 @@ func (container *Container) configureNetwork(networkName, service, networkDriver
+		return n, nil
+	}
+
+	// Find by partial network id
```
I've seen this piece of code upper in the patch =/

icecrime commented:
Done with my first pass of code review: looks good overall, just some questions here and there. Moving on with the manual testing :dancer:

mavenugo commented:
thanks @icecrime i will address your comments. mostly, I have to shuffle the code around to make sure it is more aligned to docker client and API design compared to what we were following in libnetwork. we had a strict REST policeman :police_car:  . enough said :)

icecrime commented:
```
root@5b5a0c2e238a:/go/src/github.com/docker/docker# docker network connect none lonely_kare
root@5b5a0c2e238a:/go/src/github.com/docker/docker# docker network inspect none
{
    "name": "none",
    "id": "ff28d9a02f48fdf2700b18365692017a62290a3af74d6c3e96bad50395018e34",
    "driver": "null",
    "containers": {
        "de2eab48e17d21702748e7eae61fbb08a966d04b743426721f784a567440281f": {
            "endpoint": "5f98cf3d95403d9f0543b9d6e7b13ee79a88b7339141b3d8648e98fbe91ccabd"
        }
    }
}
```

Isn't this weird? Because the end result is very different from running with `--net=none` (I still have network connectivity).

mavenugo commented:
@icecrime yes. am sure the connectivity of the container is via the other network it is already connected to. But I agree , we should restrict the `docker network connect` to the `none` network,
We already restrict it for `host` and `container` network.

mavenugo commented:
@icecrime @cpuguy83 @calavera addressed most of the comments.
moved the name, partial-id, id match support from UX to API layer (much similar to containers) and simplified the API layer quite a bit by moving the backend business logic into daemon package as suggested by @cpuguy83. That took care of most the comments.

PTAL

icecrime commented:
```
# docker network create 
docker: "network create" requires 1 argument.
See 'docker network create --help'.

Usage:  docker network create [OPTIONS] NETWORK-NAME

Creates a new network with a name specified by the user
```

It would be nice to see `-d,--driver` listed here with its possible values.

mavenugo commented:
>>  It would be nice to see -d,--driver listed here with its possible values.

Unfortunately, we cannot :-/
today, our plugin architecture depends on lazy loading of network drivers/plugins when the user actually provides the driver name during the create. plugins/drivers dont register with the daemon ahead of time for us to make use of. I would love to have that capability though. unfortunately, we dont have it atm.

icecrime commented:
I don't know if it's by design, but had a container joined two networks, then disconnect him from both:

```
# docker inspect -f '{{ .NetworkSettings.Networks }}' c5b
[]
```

Entries are lingering in the `/etc/hosts` file:

```
# docker exec -ti c5b cat /etc/hosts
172.19.0.1      c5bfbd440e46
127.0.0.1       localhost
::1     localhost ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
172.19.0.1      test
172.19.0.1      test.test
172.19.0.2      test2
172.19.0.2      test2.test
```

Still, as expected, nothing is reachable:

```
# docker exec -ti c5b ping test 
PING test (172.19.0.1): 56 data bytes
ping: sendto: Network unreachable
```

ibuildthecloud commented:
@icecrime `docker volume create -h` is the same.  But would be nice of the default value was set to `bridge` so.

```
Usage:	docker network create [OPTIONS] NETWORK-NAME

Creates a new network with a name specified by the user

  -d, --driver=bridge      Driver to manage the Network
  --help=false       Print usage
```

Or whatever the default is.

icecrime commented:
@ibuildthecloud Ah makes sense, I wasn't expecting `-h` to produce different output ;-)

mavenugo commented:
@icecrime 

> Entries are lingering in the /etc/hosts file:

Yes, it is a known backend bug and will be resolved in the next vendor-in.

icecrime commented:
Besides this small things, rest LGTM, it's really super cool!

How do you want to proceed? Update vendor-in and have me run another pass of tests?

mavenugo commented:
@icecrime we have a pending vendor-in PR : #16718 which brings in the discovery changes.
These are independent PRs and sure, we can merge that first and then test this on top of it.

ibuildthecloud commented:
Merging #16718  now would be swell :)

mavenugo commented:
@icecrime now that #16718 is merged, do you want me to rebase this ?
(they are independent changes and hence merging is not mandatory). Still, if you think that will help, I can push a rebase.

icecrime commented:
It's maybe worth it for the newcomer that want to test this PR.

Otherwise LGTM from me.

mavenugo commented:
okay. let me rebase it and let the CI take its time :)

mavenugo commented:
@icecrime rebased to the latest.

cpuguy83 commented:
When calling `docker network connect` for a container that is already connected to the given network, the  error returned is `a sandbox has already joined the endpoint`, it would be great if we can get a better error message.
It's also returning a 500 error and should be 40x.

Likewise disconnecting a network from a container that is not on that network should yield a 404(???) but is currently yielding 500 and has an error message `could not locate network endpoint for container 89900d`. Might be nice to have a nicer error there.

Trying to delete a network that is in use yields a 500, seems like it should be a 40x. Gramatically is also a bit weird `network with name test id 5c6d64aa6138823f584e6a3c8f6975323b738361e1a627e3a6f4e4a21a759434 has active endpoints`.
Maybe `network has active endpoints`, and potentially list the containers.

Creating a network that already exists yields a 500, should be a 40x. API docs only list `201` as a valid response.

Wondering if a container that has a network that no longer exists should fail to start, currently only a warning is given, and only in the daemon logs.

vieux commented:
When you disconnect a container from all it's networks, shouldn't it fallback in the none network ?

```
#1
$ docker run -d --name test ubuntu ping google.com
#2
$ docker network disconnect bridge test
#3
$ docker network connect none test
#4
```

Is there any difference between `#3` and `#4` ?

mavenugo commented:
@vieux attaching a container to none will typically add an loopback interface. but disconnecting from other existing networks might remove the loopback interface (depending on the driver being used for that network). So that is the only difference. It is a very subtle difference. BTW, in our bridge network, we don't remove the loopback interface during disconnect. So you wont find any difference between #3 and #4 if the network being disconnected is backed by the default bridge network.

mavenugo commented:
@cpuguy83 thanks for all the good comments. for your comments (https://github.com/docker/docker/pull/16645#issuecomment-145719089), it requires some backend changes to have typed and proper error messages in libnetwork. Have opened https://github.com/docker/libnetwork/issues/597 to track that change. 
Will update the PR that addresses rest of your comments.

mavenugo commented:
@cpuguy83 actually I didnt know the awesome support of https://github.com/docker/docker/blob/master/api/server/httputils/httputils.go#L138 where the http return codes are auto-generated based on the error message. FTW. unfortunately, that requires backend changes in libnetwork. Hence we will address all of your comments in https://github.com/docker/docker/pull/16645#issuecomment-145719089 as part of https://github.com/docker/libnetwork/issues/597.

cpuguy83 commented:
@mavenugo Oh please don't rely on that hack :)

mavenugo commented:
@cpuguy83 okay even without that (awesome !) hack, we need backend change to return with appropriate error types which can be matched to return proper error code.

aanand commented:
There's a regression in this branch where containers with dots in the name won't start:

```
$ docker run --name foo.bar busybox true
Error response from daemon: Cannot start container f97c55ac6d975e556fcb9ccb161fa14daf6cc7a406b3f51656bf7165c7432b4f: invalid name: foo.bar
```

This appears to originate in libnetwork (`ErrInvalidName`), but I can't tell which function is returning it without further digging.

mavenugo commented:
@aanand thanks for catching it. Yes it's a bug in libnetwork and will be resolved in the subsequent vendor in from libnetwork that will carry a few issues identified in this PR.

mavenugo commented:
@aanand https://github.com/docker/libnetwork/issues/601. thanks again.

cpuguy83 commented:
LGTM moving to docs review

cpuguy83 commented:
ping @thaJeztah @moxiegirl

mavenugo commented:
thanks @cpuguy83

@moxiegirl & @dave-tucker are already working on networking docs revamp under https://github.com/moxiegirl/docker/tree/new-networking-docs on top of the changes included in this PR. we have a networking docs plan and executing on that.

thaJeztah commented:
@mavenugo those changes will be in a separate PR?

>  actually I didnt know the awesome support of https://github.com/docker/docker/blob/master/api/server/httputils/httputils.go#L138 

😂 yes, it's a beauty!

mavenugo commented:
@thaJeztah yes. @moxiegirl can you please confirm ?

> 😂 yes, it's a beauty!

Say whatever, it is very practical. :)

moxiegirl commented:
@thaJeztah @cpuguy83 Given the amount of work involved in reworking the documentation for this.  The plan is to pull this PR for the technical alone.  I've already got a branch ready to go with the documentation changes.  @dave-tucker  and I are working in tandem as noted by @mavenugo .

LGTM for docs to move forward under this plan.

cpuguy83 commented:
Awesome, merging!

icecrime commented:
BAM @cpuguy83!

cpuguy83 commented:
That @icecrime, such a snake!

icecrime commented:
![image](https://cloud.githubusercontent.com/assets/1564054/10346285/41d38baa-6ce3-11e5-9705-34db90b20c5e.png)

mavenugo commented:
lol... thanks @icecrime @cpuguy83 
:fireworks: :fireworks: :tada:

manoj0077 commented:
@mavenugo coming to point raised by @vieux even when the container is disconnected from network and removed ...loop back interface is not getting deleted...../var/run/docker/netns is getting populated in this scenario....

guybrush commented:
i cant reproduce the example in the UX-doc with `1.9.0-rc4`, am i doing something wrong? 

`/etc/hosts` does not get updated: https://gist.github.com/guybrush/28bbf90b9c32244465f7

thaJeztah commented:
@guybrush network docs will be updated soon; updating `/etc/hosts` (service discovery) will not be done on the default "bridge" network, only with other networks (i.e. `docker network create foo` and `docker run -itd --name c1 --net foo nginx`)

guybrush commented:
woah that was fast response, i cant wait to use the new feature :D 

thanks!

manoj0077 commented:
@mavenugo:-  Can I pass multiple networks in --net option ?

thaJeztah commented:
@manoj0077 it's not supported at this moment (it would be a nice addition though), but you can dynamically add a container to a network;

    docker network connect <networkname> <containername>

For documentation related to the new networking, see https://github.com/docker/docker/blob/master/docs/userguide/networking/index.md

manoj0077 commented:
@thaJeztah So is there a possibility to add multiple interfaces at the time of starting of container itself....

mavenugo commented:
@manoj0077 no. it is not possible currently.

zrml commented:
Is this still an experimental feature in 1.10? 
Thank you

thaJeztah commented:
@zrml no, it's no longer experimental https://docs.docker.com/engine/userguide/networking/dockernetworks/
